File: app/src/main/java/com/holokenmod/ui/KeyPadFragment.java
package com.holokenmod.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TableLayout;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.holokenmod.Game;
import com.holokenmod.R;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.options.GameVariant;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class KeyPadFragment extends Fragment implements GridCreationListener {
    private final List<Button> numbers = new ArrayList<>();
    private final List<Button> allNumbers = new ArrayList<>();
    private Button numberExtra;
    private TableLayout controlKeypad;
    private Game game;
    
    public KeyPadFragment() {
        super(R.layout.key_pad_fragment);
    }
    
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
        // Defines the xml file for the fragment
        return inflater.inflate(R.layout.key_pad_fragment, parent, false);
    }
    
    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        numbers.add(view.findViewById(R.id.button1));
        numbers.add(view.findViewById(R.id.button2));
        numbers.add(view.findViewById(R.id.button3));
        numbers.add(view.findViewById(R.id.button4));
        numbers.add(view.findViewById(R.id.button5));
        numbers.add(view.findViewById(R.id.button6));
        numbers.add(view.findViewById(R.id.button7));
        numbers.add(view.findViewById(R.id.button8));
        numbers.add(view.findViewById(R.id.button9));
        numbers.add(view.findViewById(R.id.button10));
        numberExtra = view.findViewById(R.id.buttonExtra);
    
        allNumbers.addAll(numbers);
        allNumbers.add(numberExtra);
    
        this.controlKeypad = view.findViewById(R.id.controls);
    
        //this.controlKeypad.setVisibility(View.INVISIBLE);
    
        for (final Button numberButton : numbers) {
            addButtonListeners(numberButton);
        }
        addButtonListeners(numberExtra);

        if (game != null) {
            setButtonLabels();
            setButtonVisibility();
        }
    }
    
    private void addButtonListeners(Button numberButton) {
        numberButton.setOnClickListener(v -> {
            final int d = Integer.parseInt(((Button) v).getText().toString());
            game.enterPossibleNumber(d);
        });
        numberButton.setOnLongClickListener(v -> {
            final int d = Integer.parseInt(((Button) v).getText().toString());
            game.enterNumber(d);
            
            return true;
        });
    }
    
    
    public void freshGridWasCreated() {
        if (!isAdded()) {
            return;
        }
    
        requireActivity().runOnUiThread( () -> {
            setButtonLabels();
            setButtonVisibility();
        });
    }
    
    private void setButtonLabels() {
        DigitSetting digitSetting = GameVariant.getInstance().getDigitSetting();
        
        /*if (digitSetting == DigitSetting.FIRST_DIGIT_ZERO) {
            numberExtra.setText("0");
        } else {
            numberExtra.setText("11");
        }*/
        
        Iterator<Integer> digits = digitSetting.getAllNumbers().iterator();
        
        for (final Button numberButton : numbers) {
            int digit = digits.next();
            
            numberButton.setText(Integer.toString(digit));
            numberButton.setVisibility(View.VISIBLE);
        }
    }
    
    private void setButtonVisibility() {
        DigitSetting digitSetting = GameVariant.getInstance().getDigitSetting();
        
        for (int i = 0; i < numbers.size(); i++) {
            numbers.get(i).setEnabled(i < digitSetting.getMaximumDigit(game.getGrid().getGridSize()));
        }
        
        boolean useExtraNumber = digitSetting == DigitSetting.FIRST_DIGIT_ZERO
                || game.getGrid().getGridSize().getAmountOfNumbers() >= 11;
        
        numberExtra.setEnabled(useExtraNumber);
        
        this.controlKeypad.setVisibility(View.VISIBLE);
    }
    
    public void setGame(Game game) {
        this.game = game;
    
        freshGridWasCreated();
    }
}

File: gauguin-app/src/main/kotlin/org/piepmeyer/gauguin/ui/main/KeyPadFragment.kt
package org.piepmeyer.gauguin.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.button.MaterialButton
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.piepmeyer.gauguin.R
import org.piepmeyer.gauguin.game.Game
import org.piepmeyer.gauguin.ui.WindowClassCalculator
import kotlin.math.ceil
import kotlin.math.roundToInt
import kotlin.properties.Delegates

class KeyPadFragment :
    Fragment(),
    KoinComponent {
    private val game: Game by inject()

    private val numbers = mutableListOf<MaterialButton>()

    private val numberButtonToDigit = mutableMapOf<MaterialButton, Int>()

    private lateinit var layoutCalculator: KeyPadLayoutCalculator
    private var layoutId by Delegates.notNull<Int>()
    private lateinit var rootView: View

    override fun onCreateView(
        inflater: LayoutInflater,
        parent: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        layoutCalculator = KeyPadLayoutCalculator(WindowClassCalculator(requireActivity()))
        layoutId = layoutCalculator.calculateLayoutId(game.grid)

        rootView = inflater.inflate(layoutId, parent, false)

        numbers += rootView.findViewById<MaterialButton>(R.id.button1)
        numbers += rootView.findViewById<MaterialButton>(R.id.button2)
        numbers += rootView.findViewById<MaterialButton>(R.id.button3)
        numbers += rootView.findViewById<MaterialButton>(R.id.button4)
        numbers += rootView.findViewById<MaterialButton>(R.id.button5)
        numbers += rootView.findViewById<MaterialButton>(R.id.button6)
        numbers += rootView.findViewById<MaterialButton>(R.id.button7)
        numbers += rootView.findViewById<MaterialButton>(R.id.button8)
        numbers += rootView.findViewById<MaterialButton>(R.id.button9)
        numbers += rootView.findViewById<MaterialButton>(R.id.button10)
        numbers += rootView.findViewById<MaterialButton>(R.id.button11)
        numbers += rootView.findViewById<MaterialButton>(R.id.button12)

        return rootView
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        numbers.forEach {
            addButtonListeners(it)
        }

        val viewModel = ViewModelProvider(requireActivity())[MainViewModel::class.java]

        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.gameStateWithGrid.collect {
                    if (it.state == GameState.PLAYING) {
                        requireActivity().runOnUiThread {
                            if (layoutId != layoutCalculator.calculateLayoutId(game.grid)) {
                                requireActivity().recreate()
                            } else {
                                setButtonStates()
                            }
                        }
                    }

                    rootView.visibility =
                        if (it.state in listOf(GameState.PLAYING, GameState.CALCULATING_NEW_GRID)) {
                            View.VISIBLE
                        } else {
                            View.INVISIBLE
                        }
                }
            }
        }

        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.fastFinishingModeState.collect {
                    setButtonStates()
                }
            }
        }
    }

    private fun addButtonListeners(numberButton: MaterialButton) {
        numberButton.setOnClickListener {
            game.enterPossibleNumber(numberButtonToDigit.getValue(numberButton))
        }
        numberButton.setOnLongClickListener {
            game.enterNumber(numberButtonToDigit.getValue(numberButton))
            true
        }
    }

    private fun setButtonStates() {
        val visibleRows = ceil(game.grid.variant.possibleDigits.size / 3.0).toInt()
        val lastVisibleNumber = visibleRows * 3 - 1

        val digitSetting = game.grid.variant.options.digitSetting
        val digits = digitSetting.numbers.toMutableList()

        if (digitSetting.zeroOnKeyPadShouldBePlacedAtLast()) {
            digits.remove(0)
            digits.add(lastVisibleNumber, 0)
        }

        val digitsIterator = digits.iterator()

        var i = 0

        numberButtonToDigit.clear()

        numbers.forEach {
            val digit = digitsIterator.next()

            numberButtonToDigit[it] = digit

            it.text = displaybleText(digit)
            it.visibility =
                when {
                    (i <= lastVisibleNumber && !game.isInFastFinishingMode()) -> View.VISIBLE
                    (i <= lastVisibleNumber && game.isInFastFinishingMode()) -> View.INVISIBLE
                    else -> View.GONE
                }
            it.isEnabled =
                game.grid.variant.possibleDigits
                    .contains(digit)
            i++
        }

        val padding = layoutCalculator.calculateLayoutMarginBottom(game.grid)
        rootView.updateLayoutParams<ViewGroup.MarginLayoutParams> { }
        rootView.setPaddingRelative(0, padding, 0, padding)
        rootView.invalidate()
    }

    private fun displaybleText(digit: Int): String {
        var text =
            game.grid.variant.options.numeralSystem
                .displayableString(digit)

        if (text.length > 4) {
            val cutTextIndex = (text.length / 2 + 0.4).roundToInt()

            text = "${text.subSequence(0, cutTextIndex)}\n${
                text.subSequence(
                    cutTextIndex,
                    text.length,
                )
            }"
        }

        return text
    }
}
